### DISCUSSION TITLE
Notifications?

### DISCUSSION THREAD

**Discussion by IcedQuinn** [seaweedfs/seaweedfs#2447] [Ideas]
Some blob stores have notification systems (ex. send to MQTT or a webhook when things are added to buckets, deleted, expire etc.) Is there an equivalent to that?

**Comment by chrislusf** [comment]
You can send the events to Kafka https://github.com/chrislusf/seaweedfs/wiki/Async-Replication-to-Cloud

Or subscribe to the events directly:

https://github.com/chrislusf/seaweedfs/wiki/Filer-Change-Data-Capture
and
https://github.com/chrislusf/seaweedfs/wiki/Filer-Metadata-Events#subscribe-to-metadata

**Reply by IcedQuinn** [reply]
kafka is a _very_ heavy requirement compared to ex. mqtt or webhooks.

hooking filer might be fine (although i've just been testing s3 buckets, not as much the filesystem overlay) though i'll have to look at how to write a gRPC system for the languages i use 🤔

**Reply by alexellis** [reply]
👍 

I added support to Minio for webhook / HTTP notifications back in 2017. It's broadly used now, and trivial for consumers.

It looks like Kafka is the only viable and portable method. We're looking at using SeaweedFS for a migration guide from AWS Lambda to OpenFaaS with SeaweedFS.

Would you reconsider putting HTTP webhooks onto your backlog @chrislusf?

**Reply by chrislusf** [reply]
What is the design for it?

**Reply by alexellis** [reply]
Here's the design from AWS (after which SeaweedFS is modelled): https://docs.aws.amazon.com/AmazonS3/latest/userguide/EventNotifications.html

I would expect at a minimum to be able to register a HTTP endpoint, and say which events I want to get a callback for: ObjectPut, etc

Here are the docs for how Minio does it - https://min.io/docs/minio/linux/administration/monitoring/publish-events-to-webhook.html#minio-bucket-notifications-publish-webhook

And the docs for how Ceph does this:

https://docs.ceph.com/en/quincy/radosgw/notifications/

So, think of your current Kafka integration, and instead of publishing to a topic, you're making a HTTP POST to a registered endpoint.

Does this make sense now, or is there anything else I can add?

**Reply by chrislusf** [reply]
Thanks for the examples, which actually covered event types, how to configure, and how to process them.

For SeaweedFS, will it be sync or async notifications? How to handle errors?

SeaweedFS has grpc api to listen for metadata updates? Will it be sufficient?
